Move MediaRecorderPrivateWriterAVFObjC to GPU process

jyavenard opened the pull request:
#### 34b2d01aae84eaca2ab099ff4d1cc5b6e89a6b69
<pre>
Move MediaRecorderPrivateWriterAVFObjC to GPU process
<a href="https://bugs.webkit.org/show_bug.cgi?id=284705">https://bugs.webkit.org/show_bug.cgi?id=284705</a>
<a href="https://rdar.apple.com/141503370">rdar://141503370</a>

Reviewed by Youenn Fablet.

On iOS, the AVAssetWriter can&apos;t run in the Web Process as it is blocked by the sandbox.
We need to move it to the more priviledged GPU process.

We add a MediaStrategy createMediaRecorderPrivateWriter which will return a
RemoteMediaRecorderPrivateWriter.
MediaRecorderPrivateWriter requests will be handled by the RemoteMediaRecorderPrivateWriterManager
which keeps a map of RemoteMediaRecorderPrivateWriterProxy the communictates with their
MediaRecorderPrivateWriterAVFObjC.
The muxed data is then returned to the content proceess&apos;s MediaRecorder object.

Communication between the RemoteMediaRecorderPrivateWriter and its RemoteMediaRecorderPrivateWriterProxy
is asynchronous except on setup where we use a sync call to create the tracks.

No change in observable behaviours. Covered by existing tests.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/MediaStrategy.cpp:
(WebCore::MediaStrategy::createMediaRecorderPrivateWriter const):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::create):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::GPUConnectionToWebProcess::remoteMediaRecorderPrivateWriterManager):
(WebKit::GPUConnectionToWebProcess::protectedRemoteMediaRecorderPrivateWriterManager):
(WebKit::GPUConnectionToWebProcess::dispatchMessage):
(WebKit::GPUConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.cpp: Added.
(WebKit::RemoteMediaRecorderPrivateWriterProxy::create):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::addAudioTrack):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::addVideoTrack):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::allTracksAdded):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::writeFrames):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::close):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::takeData):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::RemoteMediaRecorderPrivateWriterProxy):
(WebKit::RemoteMediaRecorderPrivateWriterManager::RemoteMediaRecorderPrivateWriterManager):
(WebKit::RemoteMediaRecorderPrivateWriterManager::ref const):
(WebKit::RemoteMediaRecorderPrivateWriterManager::deref const):
(WebKit::RemoteMediaRecorderPrivateWriterManager::create):
(WebKit::RemoteMediaRecorderPrivateWriterManager::addAudioTrack):
(WebKit::RemoteMediaRecorderPrivateWriterManager::addVideoTrack):
(WebKit::RemoteMediaRecorderPrivateWriterManager::allTracksAdded):
(WebKit::RemoteMediaRecorderPrivateWriterManager::writeFrames):
(WebKit::RemoteMediaRecorderPrivateWriterManager::close):
(WebKit::RemoteMediaRecorderPrivateWriterManager::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.h: Added.
(WebKit::RemoteMediaRecorderPrivateWriterManager::allowsExitUnderMemoryPressure):
* Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.messages.in: Added.
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.h: Added.
(WebKit::RemoteAudioInfo::RemoteAudioInfo):
(WebKit::RemoteAudioInfo::toAudioInfo const):
(WebKit::RemoteVideoInfo::RemoteVideoInfo):
(WebKit::RemoteVideoInfo::toVideoInfo const):
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.serialization.in: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(conditions_for_header):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaRecorderPrivateWriter.cpp: Added.
(WebKit::RemoteMediaRecorderPrivateWriter::create):
(WebKit::RemoteMediaRecorderPrivateWriter::RemoteMediaRecorderPrivateWriter):
(WebKit::RemoteMediaRecorderPrivateWriter::addAudioTrack):
(WebKit::RemoteMediaRecorderPrivateWriter::addVideoTrack):
(WebKit::RemoteMediaRecorderPrivateWriter::allTracksAdded):
(WebKit::RemoteMediaRecorderPrivateWriter::writeFrames):
(WebKit::RemoteMediaRecorderPrivateWriter::writeFrame):
(WebKit::RemoteMediaRecorderPrivateWriter::close):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaRecorderPrivateWriter.h: Added.
* Source/WebKit/WebProcess/GPU/media/RemoteMediaRecorderPrivateWriterIdentifier.h: Copied from Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h.
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createMediaRecorderPrivateWriter const):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:

Canonical link: <a href="https://commits.webkit.org/288007@main">https://commits.webkit.org/288007@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e485ac0f6897982699ef7a310a52fc294af1728

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81672 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1197 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35621 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86212 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32662 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83778 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1228 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9016 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63711 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21443 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/876 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44003 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/81125 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28497 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31117 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29095 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87651 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8909 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72049 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9092 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70145 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71284 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17747 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15364 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14278 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8861 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8701 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12222 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10510 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
```diff
@@ -43,6 +43,11 @@ MediaRecorderPrivateWriter::~MediaRecorderPrivateWriter() = default;
 {
+    if (hasPlatformStrategies()) {
+        auto writer = platformStrategies()->mediaStrategy().createMediaRecorderPrivateWriter(type, listener);
+        if (writer)
```
```suggestion
        if (auto writer = platformStrategies()->mediaStrategy().createMediaRecorderPrivateWriter(type, listener))
```

youennf commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -166,6 +169,10 @@ class GPUConnectionToWebProcess
 #endif
+#if PLATFORM(COCOA) && ENABLE(MEDIA_RECORDER)
+    RemoteMediaRecorderPrivateWriterManager& remoteMediaRecorderPrivateWriterManager();
+        Ref<RemoteMediaRecorderPrivateWriterManager> protectedRemoteMediaRecorderPrivateWriterManager();
```
```suggestion
    Ref<RemoteMediaRecorderPrivateWriterManager> protectedRemoteMediaRecorderPrivateWriterManager();
```

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.cpp:
```diff
@@ -0,0 +1,193 @@
+    Ref<SharedBuffer> takeData() { return m_data.takeAsContiguous(); }
+
+private:
+    RemoteMediaRecorderPrivateWriterProxy(const String& mimeType)
```
explicit

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.cpp:
```diff
@@ -0,0 +1,193 @@
+
+void RemoteMediaRecorderPrivateWriterManager::addAudioTrack(RemoteMediaRecorderPrivateWriterIdentifier identifier, RemoteAudioInfo info, CompletionHandler<void(std::optional<uint8_t>)>&& completionHandler)
+{
+    ASSERT(m_remoteMediaRecorderPrivateWriters.contains(identifier));
```
```suggestion
   if (!m_remoteMediaRecorderPrivateWriters.contains(identifier))
       return;
```

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.cpp:
```diff
@@ -0,0 +1,193 @@
+
+void RemoteMediaRecorderPrivateWriterManager::addVideoTrack(RemoteMediaRecorderPrivateWriterIdentifier identifier, RemoteVideoInfo info, std::optional<CGAffineTransform> transform, CompletionHandler<void(std::optional<uint8_t>)>&& completionHandler)
+{
+    ASSERT(m_remoteMediaRecorderPrivateWriters.contains(identifier));
```
Ditto here and below.

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.cpp:
```diff
@@ -0,0 +1,193 @@
+
+    Deque<UniqueRef<MediaSamplesBlock>> samples;
+    for (auto& sample : vectorSamples)
+        samples.append(makeUniqueRef<MediaSamplesBlock>(sample.first == TrackInfo::TrackType::Audio ? m_audioInfo.get() : static_cast<TrackInfo*>(m_videoInfo.get()), WTFMove(sample.second)));
```
Deque<UniqueRef<>> does not seem very well suited here.

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.h:
```diff
@@ -0,0 +1,99 @@
+    WTF_MAKE_TZONE_ALLOCATED(RemoteMediaRecorderPrivateWriterManager);
+    WTF_MAKE_NONCOPYABLE(RemoteMediaRecorderPrivateWriterManager);
+public:
+    RemoteMediaRecorderPrivateWriterManager(GPUConnectionToWebProcess&);
```
explicit

youennf commented on Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.h:
```diff
@@ -0,0 +1,99 @@
+
+    bool allowsExitUnderMemoryPressure() { return m_remoteMediaRecorderPrivateWriters.isEmpty(); }
+
+    // Messages.
```
All these could probably be private.

youennf commented on Source/WebKit/GPUProcess/media/RemoteTrackInfo.h:
```diff
@@ -0,0 +1,135 @@
+    {
+    }
+    // Used by IPC generator
+    RemoteAudioInfo(WebCore::FourCC codecName, const String& codecString, WebCore::TrackID trackID, uint32_t rate, uint32_t channels, uint32_t framesPerPacket, uint8_t bitDepth, RefPtr<WebCore::SharedBuffer> cookieData)
```
We could probably use a String&& here and move codecString

youennf commented on Source/WebKit/GPUProcess/media/RemoteTrackInfo.h:
```diff
@@ -0,0 +1,135 @@
+    {
+    }
+    // Used by IPC generator
+    RemoteVideoInfo(WebCore::FourCC codecName, const String& codecString, WebCore::TrackID trackID, WebCore::FloatSize size, WebCore::FloatSize displaySize, uint8_t bitDepth, WebCore::PlatformVideoColorSpace colorSpace, RefPtr<WebCore::SharedBuffer> atomData)
```
Ditto

youennf commented on Source/WebKit/GPUProcess/media/RemoteTrackInfo.serialization.in:
```diff
@@ -0,0 +1,49 @@
+#if ENABLE(GPU_PROCESS) && ENABLE(VIDEO)
+
+header: "RemoteTrackInfo.h"
+[CustomHeader] struct WebKit::RemoteAudioInfo {
```
We could probably use that pattern for MediaSampleBlock as well.

jyavenard committed:
Move MediaRecorderPrivateWriterAVFObjC to GPU process
https://bugs.webkit.org/show_bug.cgi?id=284705
rdar://141503370

Reviewed by Youenn Fablet.

On iOS, the AVAssetWriter can't run in the Web Process as it is blocked by the sandbox.
We need to move it to the more priviledged GPU process.

We add a MediaStrategy createMediaRecorderPrivateWriter which will return a
RemoteMediaRecorderPrivateWriter.
MediaRecorderPrivateWriter requests will be handled by the RemoteMediaRecorderPrivateWriterManager
which keeps a map of RemoteMediaRecorderPrivateWriterProxy the communictates with their
MediaRecorderPrivateWriterAVFObjC.
The muxed data is then returned to the content proceess's MediaRecorder object.

Communication between the RemoteMediaRecorderPrivateWriter and its RemoteMediaRecorderPrivateWriterProxy
is asynchronous except on setup where we use a sync call to create the tracks.

No change in observable behaviours. Covered by existing tests.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/MediaStrategy.cpp:
(WebCore::MediaStrategy::createMediaRecorderPrivateWriter const):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::create):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::allowsExitUnderMemoryPressure const):
(WebKit::GPUConnectionToWebProcess::remoteMediaRecorderPrivateWriterManager):
(WebKit::GPUConnectionToWebProcess::protectedRemoteMediaRecorderPrivateWriterManager):
(WebKit::GPUConnectionToWebProcess::dispatchMessage):
(WebKit::GPUConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.cpp: Added.
(WebKit::RemoteMediaRecorderPrivateWriterProxy::create):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::addAudioTrack):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::addVideoTrack):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::allTracksAdded):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::writeFrames):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::close):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::takeData):
(WebKit::RemoteMediaRecorderPrivateWriterProxy::RemoteMediaRecorderPrivateWriterProxy):
(WebKit::RemoteMediaRecorderPrivateWriterManager::RemoteMediaRecorderPrivateWriterManager):
(WebKit::RemoteMediaRecorderPrivateWriterManager::ref const):
(WebKit::RemoteMediaRecorderPrivateWriterManager::deref const):
(WebKit::RemoteMediaRecorderPrivateWriterManager::create):
(WebKit::RemoteMediaRecorderPrivateWriterManager::addAudioTrack):
(WebKit::RemoteMediaRecorderPrivateWriterManager::addVideoTrack):
(WebKit::RemoteMediaRecorderPrivateWriterManager::allTracksAdded):
(WebKit::RemoteMediaRecorderPrivateWriterManager::writeFrames):
(WebKit::RemoteMediaRecorderPrivateWriterManager::close):
(WebKit::RemoteMediaRecorderPrivateWriterManager::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.h: Added.
(WebKit::RemoteMediaRecorderPrivateWriterManager::allowsExitUnderMemoryPressure):
* Source/WebKit/GPUProcess/media/RemoteMediaRecorderPrivateWriterManager.messages.in: Added.
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.h: Added.
(WebKit::RemoteAudioInfo::RemoteAudioInfo):
(WebKit::RemoteAudioInfo::toAudioInfo const):
(WebKit::RemoteVideoInfo::RemoteVideoInfo):
(WebKit::RemoteVideoInfo::toVideoInfo const):
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.serialization.in: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(conditions_for_header):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaRecorderPrivateWriter.cpp: Added.
(WebKit::RemoteMediaRecorderPrivateWriter::create):
(WebKit::RemoteMediaRecorderPrivateWriter::RemoteMediaRecorderPrivateWriter):
(WebKit::RemoteMediaRecorderPrivateWriter::addAudioTrack):
(WebKit::RemoteMediaRecorderPrivateWriter::addVideoTrack):
(WebKit::RemoteMediaRecorderPrivateWriter::allTracksAdded):
(WebKit::RemoteMediaRecorderPrivateWriter::writeFrames):
(WebKit::RemoteMediaRecorderPrivateWriter::writeFrame):
(WebKit::RemoteMediaRecorderPrivateWriter::close):
* Source/WebKit/WebProcess/GPU/media/RemoteMediaRecorderPrivateWriter.h: Added.
* Source/WebKit/WebProcess/GPU/media/RemoteMediaRecorderPrivateWriterIdentifier.h: Copied from Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h.
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createMediaRecorderPrivateWriter const):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:

Canonical link: https://commits.webkit.org/288007@main
